common_code/solver_cg_pipelined.h

```diff
@@ -0,0 +1,273 @@
+     const dealii::ArrayView<T> &      sums)
+{
+  iall_reduce(MPI_SUM, values, mpi_communicator, mpi_request, sums);
+}
+
+
+template <typename VectorType>
+class SolverCGPipelined : public dealii::SolverBase<VectorType>
+{
+public:
+  /**
+   * Declare type for container size.
+   */
+  using size_type = dealii::types::global_dof_index;
+
+  /**
+   * Constructor.
+   */
+  SolverCGPipelined(dealii::SolverControl &cn)
+    : dealii::SolverBase<VectorType>(cn)
+  {}
+
+
+  /**
+   * Virtual destructor.
+   */
+  virtual ~SolverCGPipelined() override = default;
+
+  /**
+   * Solve the linear system $Ax=b$ for x.
+   */
+
+  template <typename MatrixType, typename PreconditionerType>
+  void
+  solve(const MatrixType &A, VectorType &x, const VectorType &b, const PreconditionerType &)
+
+  {
+    if (std::is_same<PreconditionerType, dealii::PreconditionIdentity>::value == false)
+      {
+        Assert(true, ExcNotImplemented());
+      }
+
+    using number                      = typename VectorType::value_type;
+    dealii::SolverControl::State conv = dealii::SolverControl::iterate;
+
+    // Memory allocation
+    typename dealii::VectorMemory<VectorType>::Pointer z_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer s_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer p_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer r_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer w_pointer(this->memory);
+    typename dealii::VectorMemory<VectorType>::Pointer q_pointer(this->memory);
+
+    // some aliases for simpler access
+    VectorType &z = *z_pointer;
+    VectorType &s = *s_pointer;
+    VectorType &p = *p_pointer;
+    VectorType &r = *r_pointer;
+    VectorType &w = *w_pointer;
+    VectorType &q = *q_pointer;
+
+    int    it  = 0;
+    double res = std::numeric_limits<double>::max();
+
+    // resize the vectors, but do not set
+    // the values since they'd be overwritten
+    // soon anyway.
+    z.reinit(x, false);
+    s.reinit(x, false);
+    p.reinit(x, false);
+
+    r.reinit(x, true);
+    w.reinit(x, true);
+    q.reinit(x, true);
+
+    number beta      = 0.;
+    number gamma     = 0.;
+    number alpha     = 0.;
+    number gamma_old = 0.;
+    number delta     = 0.;
+
+    // compute residual. if vector is
+    // zero, then short-circuit the
+    // full computation
+    if (!x.all_zero())
+      {
+        A.vmult(r, x);
+        r.add(-1., b);
+      }
+    else
+      r.equ(-1., b);
+
+    r *= -1.0;
+
+    res = r.l2_norm();
+
+    conv = this->iteration_status(0, res, x);
+    if (conv != dealii::SolverControl::iterate)
+      return;
+
+    A.vmult(w, r);
+
+    NonBlockingDotproduct<number> nonblocking;
+
+    while (conv == dealii::SolverControl::iterate)
+      {
+        it++;
+
+        nonblocking.dot_product_start(r, r, &gamma);
+        nonblocking.dot_product_start(w, r, &delta);
```

Comment:
I guess these two could be merge so that there is only a single read access to `r`.